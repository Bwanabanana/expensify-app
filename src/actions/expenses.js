import uuid from 'uuid';
import { Auth } from 'aws-amplify';

export const addExpenseAction = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const addExpense = (expense = {}) => {
    return (dispatch) => {
        const {
            id = uuid(),
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expense;

        return Auth.currentSession().then((res) => {
            const cognitoToken = res.getIdToken();
            const jwt = cognitoToken.getJwtToken();
            return jwt;
        }).then((jwt) => {
            const url = `https://api.expensify.bwanabanana.com/expenses/${id}`;
            const options = {
                method: 'PUT',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': jwt
                },
                body: JSON.stringify({ description, note, amount, createdAt })
            };
            return fetch(url, options);
        }).then((response) => {
            dispatch(addExpenseAction({ id, description, note, amount, createdAt }));
        });
    };
};

export const removeExpenseAction = (id) => ({
    type: 'REMOVE_EXPENSE',
    id //shorthand for id: id
});

export const removeExpense = (id) => {
    return (dispatch) => {

        return Auth.currentSession().then((res) => {
            const cognitoToken = res.getIdToken();
            const jwt = cognitoToken.getJwtToken();
            return jwt;
        }).then((jwt) => {
            const url = `https://api.expensify.bwanabanana.com/expenses/${id}`;
            const options = {
                method: 'DELETE',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': jwt
                }
            };
            return fetch(url, options);
        }).then((response) => {
            dispatch(removeExpenseAction(id));
        });
    };
};

export const editExpenseAction = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

export const editExpense = (id, updates) => {

    return (dispatch) => {

        return Auth.currentSession().then((res) => {
            const cognitoToken = res.getIdToken();
            const jwt = cognitoToken.getJwtToken();
            return jwt;
        }).then((jwt) => {
            const {
                description = '',
                note = '',
                amount = 0,
                createdAt = 0
            } = updates;

            const url = `https://api.expensify.bwanabanana.com/expenses/${id}`;
            const options = {
                method: 'PUT',
                headers: {
                    'Content-Type': 'application/json',
                    'Authorization': jwt
                },
                body: JSON.stringify({ description, note, amount, createdAt })
            };
            return fetch(url, options);
        }).then((response) => {
            dispatch(editExpenseAction(id, updates));
        });
    };
};

export const setExpensesAction = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const loadExpenses = () => {
    return (dispatch) => {

        return Auth.currentSession().then((res) => {
            const cognitoToken = res.getIdToken();
            const jwt = cognitoToken.getJwtToken();
            return jwt;
        }).then((jwt) => {
            const url = 'https://api.expensify.bwanabanana.com/expenses';
            const options = {
                headers: {
                    Authorization: jwt
                }
            };
            return fetch(url, options);
        }).then((response) => {
            return response.json();
        }).then((json) => {
            console.log(json);
            dispatch(setExpensesAction(json.expenses));
        });
    };
};


