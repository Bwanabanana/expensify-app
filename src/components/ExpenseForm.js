import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates-temp';

export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({ error: 'Please provide description and amount' }));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };
    render() {
        return (
            <form
                className="form"
                onSubmit={this.onSubmit}
            >
                {this.state.error && <p className="form__error">{this.state.error}</p>}
                <input
                    autoFocus
                    className="text-input"
                    onChange={this.onDescriptionChange}
                    placeholder="Description"
                    text="text"
                    value={this.state.description}
                />
                <input
                    className="text-input"
                    onChange={this.onAmountChange}
                    placeholder="Amount"
                    type="text"
                    value={this.state.amount}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    displayFormat="DD/MM/YYYY"
                    focused={this.state.calendarFocused}
                    isOutsideRange={(day) => false}
                    numberOfMonths={1}
                    onDateChange={this.onDateChange}
                    onFocusChange={this.onFocusChange}
                />
                <textarea
                    className="textarea"
                    onChange={this.onNoteChange}
                    placeholder="Add a note for your expense (optional)"
                    value={this.state.note}
                />
                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        );
    }
}
