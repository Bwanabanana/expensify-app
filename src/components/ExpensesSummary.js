import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom'
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ numExpenses, expensesTotal }) => {
    const expenseWord = numExpenses == 1 ? 'expense' : 'expenses'
    const formattedExpensesTotal = numeral(expensesTotal / 100).format('Â£0,0.00');
    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">Viewing <span>{numExpenses}</span> {expenseWord} totalling <span>{formattedExpensesTotal}</span></h1>
                <div className="page-header__actions">
                    <Link className="button" to="/create">Add Expense</Link>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state, props) => {
    const expenses = selectExpenses(state.expenses, state.filters);
    return {
        expensesTotal: selectExpensesTotal(expenses),
        numExpenses: expenses.length
    };
}

export default connect(mapStateToProps)(ExpensesSummary);