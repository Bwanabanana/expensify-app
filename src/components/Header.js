import React from 'react';
import { Link } from 'react-router-dom';
import { AmplifySignOut } from '@aws-amplify/ui-react';

const Header = () => (
    <header className="header">
        <div className="content-container">
            <div className="header__content">
                <h1 className="header__title">Expensify</h1>
                <Link className="header__title" to="/">
                    <h3>Home</h3>
                </Link>
                <Link className="header__title" to="/api">
                    <h3>API</h3>
                </Link>
                <Link className="header__title" to="/help">
                    <h3>Help</h3>
                </Link>
                <AmplifySignOut />
            </div>
        </div>
    </header>
);

export default Header;