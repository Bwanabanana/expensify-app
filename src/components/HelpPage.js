import React from 'react';

const HelpPage = () => (
    <div>
        <p>No help found :(</p>
    </div>
);

export default HelpPage;