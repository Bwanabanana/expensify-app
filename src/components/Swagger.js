import React from 'react';
import PropTypes from 'prop-types';
import { Auth } from 'aws-amplify';
import SwaggerUI, { presets } from 'swagger-ui';
import 'swagger-ui/dist/swagger-ui.css';

import spec from 'js-yaml-loader!../openapi/openapi.yml';

class Swagger extends React.Component {

    componentDidMount() {
        Auth.currentSession().then((res) => {
            const cognitoToken = res.getIdToken();
            const key = cognitoToken.getJwtToken();

            const ui = SwaggerUI({
                dom_id: '#swaggerContainer',
                spec: spec,
                presets: [presets.apis],
                onComplete: () => {
                    ui.preauthorizeApiKey("expensify-CognitoUserPoolAuthorizer", key);
                }
            });
        })
    }

    render() {
        return (
            <div id="swaggerContainer" />
        );
    }
}

Swagger.propTypes = {
    url: PropTypes.string,
    spec: PropTypes.string
};

export default Swagger;