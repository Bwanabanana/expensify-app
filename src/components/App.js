import React from 'react';
import { Provider } from 'react-redux';
import Amplify from 'aws-amplify';
import AppRouter from '../routers/AppRouter';
import { withAuthenticator } from '@aws-amplify/ui-react';
import configureStore from '../store/configureStore';
import { loadExpenses } from '../actions/expenses';

const ampConfig = Amplify.configure({
    Auth: {
        region: 'eu-west-2',
        userPoolId: 'eu-west-2_59mzxsHwL',
        userPoolWebClientId: '2dmom3jhfm11466uvk7lmp91br'
    }
});

const myTheme = {
    container: {
        'alignItems': 'center',
        'background': `url('images/bg.jpg')`,
        'backgroundSize': 'cover',
        'display': 'flex',
        'height': '100vh',
        'justifyContent': 'center',
        'width': '100vw'
    }
};

const store = configureStore();

class App extends React.Component {

    constructor(props) {
        super(props);
        this.state = {
            loading: true
        };
    }

    componentDidMount() {
        store.dispatch(loadExpenses()).then(() => {
            this.setState({ loading: false });
        });
    }

    componentDidCatchError(error, info) {
        console.log('Stack', info.componentStack);
        console.log('Error', error);
    }

    render() {
        if (this.state.loading) {
            return (
                <h2>Retrieving Expenses...</h2>
            );
        } else {
            return (
                <Provider store={store}>
                    <AppRouter />
                </Provider>
            )
        }
    }
}

export default withAuthenticator(App, { includeGreetings: false }, undefined, undefined, myTheme);
