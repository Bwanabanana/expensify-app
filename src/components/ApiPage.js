import React from 'react';
import Swagger from './Swagger';

const ApiPage = () => (
    <Swagger />
);

export default ApiPage;