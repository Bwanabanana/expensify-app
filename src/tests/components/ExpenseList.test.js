import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { ExpenseList } from '../../components/ExpenseList';

test('should render expense list with expenses', () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />); // from expenses test data
    expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseList with empty message', () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});