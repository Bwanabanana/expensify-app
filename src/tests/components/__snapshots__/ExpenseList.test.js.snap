// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`should render expense list with expenses 1`] = `
<div>
  <Component
    amount={3000}
    createdAt={0}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <Component
    amount={1000}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <Component
    amount={2000}
    createdAt={345600000}
    description="Credit Card"
    id="3"
    key="3"
    note=""
  />
</div>
`;
